api/api.go

```diff
@@ -57,6 +70,63 @@ func (a *API) GetOdaoNodes(ctx context.Context, request *pb.OdaoNodesRequest) (*
 	return out, nil
 }
 
+func (a *API) GetSoloValidators(ctx context.Context, request *pb.SoloValidatorsRequest) (*pb.SoloValidators, error) {
+	a.soloValidatorCacheLock.RLock()
+	defer a.soloValidatorCacheLock.RUnlock()
+
+	out := &pb.SoloValidators{}
+	out.WithdrawalAddresses = make([][]byte, 0, len(a.soloValidatorCache))
+
+	for addr := range a.soloValidatorCache {
+		out.WithdrawalAddresses = append(out.WithdrawalAddresses, addr.Bytes())
+	}
+
+	a.m.Counter("get_solo_validators_ok").Inc()
+	return out, nil
+}
+
+func (a *API) updateCache() error {
+	a.soloValidatorCacheLock.Lock()
+	defer a.soloValidatorCacheLock.Unlock()
+
+	validators, err := a.CL.GetValidators()
+	if err != nil {
+		a.m.Counter("solo_validator_cache_update_failed").Inc()
+		return err
+	}
+
+	newMap := make(map[common.Address]interface{})
+	for _, validator := range validators {
+		if validator.Status == apiv1.ValidatorStateUnknown ||
+			validator.Status == apiv1.ValidatorStateExitedUnslashed ||
+			validator.Status == apiv1.ValidatorStateExitedSlashed ||
+			validator.Status == apiv1.ValidatorStateWithdrawalPossible ||
+			validator.Status == apiv1.ValidatorStateWithdrawalDone {
+			continue
+		}
```

Comment:
Classic Patches :) jk

In this case, I wonder if using a "whitelist" approach (i.e., looking for the statuses we care about) would be more robust, considering potential future changes to the spec. I understand the same argument can be made to the contrary (a new valid state could be added, and we would miss it). I'm sure you debated this with yourself. I'm fine either way. Just wanted to leave the note.

Reply:
With the blacklist, the risk of being wrong is that we cache a validator that isn't actually active, and give someone a token.

With a whitelist, the risk inverts, and we might omit a validator that is actually active, and refuse to give someone a token.

I figured the latter was more annoying.